src/fluree/db/ledger/indexing.clj

```diff
@@ -398,53 +400,134 @@
+        novelty-size (get-in db [:novelty :size])]
+    (> novelty-size min)))
+
+
+(defn novelty-max?
+  "Returns true if ledger is beyond novelty-max threshold."
+  [session db]
+  (let [[_ max] (novelty-min-max session)
+        novelty-size (get-in db [:novelty :size])]
+    (> novelty-size max)))
+
+
+(defn reindexed-db
+  "If a db is being reindexed, this will return the db if it is complete, else nil."
+  [session]
+  (some-> (session/indexing-promise-ch session)
+          async/poll!))
+
+
+(defn do-index
+  "Performs an index operation and returns a promise-channel of the latest db once complete"
+  [session db remove-preds]
+  (let [[lock? pc] (session/acquire-indexing-lock! session (async/promise-chan))]
+    (when lock?
+      ;; when we have a lock, reindex and put updated db onto pc.
+      (async/go
+        (let [indexed-db (<? (index db {:remove-preds remove-preds}))]
+          (<? (txproto/write-index-point-async (-> db :conn :group) indexed-db))
+          ;; updated-db might have had additional block(s) written to it, so instead
+          ;; of using it, reload from disk.
+          (session/clear-db! session)                       ;; clear db cache to force reload
+          (async/put! pc indexed-db))))
+    pc))
+
+
+(defn merge-index-block
+  "Attempts to merge newly completed index job with block-map.
+  Optimistically the new index will be one block behind the block-map,
+  but if not we will pause momentarily to try to see if things catch up
+  before throwing and error."
+  [session indexed-db {:keys [block flakes] :as block-map}]
+  (go-try
+    (let [indexed-block (:block indexed-db)
+          current?      (= indexed-block block)]
+      (log/warn "-------- merge-index-block: " current? indexed-block)
+      (if current?
+        (do
+          (session/release-indexing-lock! session indexed-block)
+          (assoc block-map :db-after indexed-db))
+        (let [_         (log/warn "^^^^^^ PAUSING  ^^^^ merge-index-block: " indexed-block block)
+              _         (async/<! (async/timeout 2000))
+              latest-db (<? (session/current-db session))
+              current?  (= (:block latest-db) block)]
+          (if current?
+            (do
+              (session/release-indexing-lock! session indexed-block)
+              (assoc block-map :db-after indexed-db))
+            (throw (ex-info (str "Recently re-indexed DB on disk is not current after 2 second pause. "
+                                 "Pending new block: " block ", latest block on disk: " (:block latest-db) ".")
+                            {:status 500 :error :db/unexpected-error}))))))))
+
+
+(defn merge-new-index
+  ([session block-map reindexed-db]
+   (merge-new-index session block-map reindexed-db 0))
+  ([session {:keys [flakes block] :as block-map} reindexed-db retries]
+   (log/warn "MERGING NEW INDEX: " {:old (:db-after block-map)
+                                    :new reindexed-db})
```

Comment:
Still want this in here? At `warn` level? With shoutiness? ;)

Reply:
Fixed logging I left in by mistake.